From f4e3028534356295faac541f7ed833cf3a4a96be Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Fri, 31 Aug 2018 15:49:36 +0300
Subject: [PATCH 4510/5242] MLK-19435 soc: imx8: pm-domains: Report failures
 to pm core

commit  3d81ae2bb96a5c3c74a90982f872885d05f13b2c from
https://source.codeaurora.org/external/imx/linux-imx.git

Right now power_on always returns true even if SCFW reports a failure.
Since the target resource is still unpowered this quickly turns into a
hang when we attempt to access it.

Handle this by reporting an error to the PM core instead and also print
the sc_err number to help with debugging.

This fixes boot on 8qm A0: instead of hanging on boot we print an error
and refuse to probe the VPU.

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Reviewed-by: Dong Aisheng <Aisheng.dong@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/pm-domains.c |   11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/soc/imx/pm-domains.c b/drivers/soc/imx/pm-domains.c
index bcb3632..a3910ca 100644
--- a/drivers/soc/imx/pm-domains.c
+++ b/drivers/soc/imx/pm-domains.c
@@ -84,8 +84,11 @@ static int imx8_pd_power(struct generic_pm_domain *domain, bool power_on)
 	sci_err = sc_pm_set_resource_power_mode(pm_ipc_handle, pd->rsrc_id,
 		(power_on) ? SC_PM_PW_MODE_ON :
 		pd->pd.state_idx ? SC_PM_PW_MODE_OFF : SC_PM_PW_MODE_LP);
-	if (sci_err)
-		pr_err("Failed power operation on resource %d\n", pd->rsrc_id);
+	if (sci_err) {
+		pr_err("Failed power operation on resource %d sc_err %d\n",
+				pd->rsrc_id, sci_err);
+		return -EINVAL;
+	}
 
 	return 0;
 }
@@ -99,6 +102,8 @@ static int imx8_pd_power_on(struct generic_pm_domain *domain)
 	pd = container_of(domain, struct imx8_pm_domain, pd);
 
 	ret = imx8_pd_power(domain, true);
+	if (ret)
+		return ret;
 
 	if (!list_empty(&pd->clks) && (pd->pd.state_idx == PD_OFF)) {
 
@@ -172,7 +177,7 @@ static int imx8_pd_power_on(struct generic_pm_domain *domain)
 		}
 	}
 
-	return ret;
+	return 0;
 }
 
 static int imx8_pd_power_off(struct generic_pm_domain *domain)
-- 
1.7.9.5

